documentation/modules/ROOT/pages/03-demo.adoc

```diff
@@ -7,28 +7,879 @@
+====
+
+==== Bundling of different logical domains
+
+Our example is also very chatty, and hard to understand at first glance.
+In fact, this is quite common in similar applications.
+The code starts small, easy to understand.
+When new features are added, it keeps growing as developers cramp new instructions into methods like `+completeTransit+`.
+
+Of course, the developers could re-architect the code, to extract instructions to a separate blocks, but that is just a half-measure.
+Still, the application will do all the operations, starting from `+completeTransit+` method in the same time, and withing the same _"script"_.
+
+=== Refactoring
+
+In this section, we'll refactor the Cabs application.
+The refactoring will be limited to make the process easy to understand.
+The scope of the refactoring will be the extraction of drivers module, which is already a separate domain in the codebase.
+Within the scope of the `+completeTransit+` method, we'll need to shift the way we calculate the fee for the driver.
+The calculation will be done asynchronously, and when the driver module publishes the calculation result, it will be saved back into the database.
+
+The base for the refactoring is the _Event Mesh_ pattern, and the asynchronous communication will be done with _Cloud Events_.
+
+==== Drivers module
+
+The functionality around drivers is already quite separated in the codebase, so it is a good staring point to extract into a separate module.
+The drivers module will become a standalone web service, deployed on the _Kubernetes_ cluster.
+The implementation of the drivers module will be done with _Rust_ for this example.
+
+Here's the _Rust_ code for calculate fee functionality.
+The entrypoint is the _Cloud Event_ of type `cabs.drivers.calculate-fee` we are expecting the _Event Mesh_ will route.
+
+[source,rust]
+----
+impl Service {
+  pub async fn calculate_fee(&mut self, ce: Event) -> Result<()> {
+    let fee_event = Self::parse_fee_event(ce)?; // <1>
+    let subject = fee_event.id.clone();
+
+    let drv = self.repo.get(&fee_event.entity.driver_id).await?;
+
+    let fee = drv.calculate_fee(&fee_event.entity.transit_price); // <2>
+
+    let fee_event = DriverFeeEvent {
+        driver_id: fee_event.entity.driver_id,
+        fee,
+    }; // <3>
+
+    let mut builder = fee_event.to_builder(); // <3>
+    if let Some(id) = subject {
+        builder = builder.subject(id);
+    } // <3>
+    let ce = builder.build().map_err(error::ErrorInternalServerError)?; // <3>
+
+    Sender::new(&self.config).send(ce).await?; // <4>
+
+    Ok(())
+  }
+  // [..]
+}
+----
+
+In the above code, we are doing the following:
+
+<1> We are parsing the internal, business logic, fee event from the _Cloud Events_ envelope.
+<2> We are calculating the fee for this event, using some business logic.
+<3> We are wrapping the calculated fee into the _Cloud Events_ envelope.
```

Comment:
into a new CloudEvent ? like with a different type?

imo is better to return different types than receiving

Reply:
It is a different type of event, although maybe the names may be a bit more distinctly sounding. We are focusing on two event types in the example:

- CalculateFee - a command event, to calculate fee for the driver for given ride
- FeeCalculated - an information event, fired when the calculator does the logic to calculate the requested fee.

You have that captured in [the conclusion diagram](https://cardil.github.io/solution-pattern-event-mesh-for-microservices/pull-1/solution-pattern-event-mesh-for-microservices/03-demo.html#_conclusion).